jax/experimental/host_callback.py

```diff
@@ -307,20 +307,28 @@ def pp_val(arg) -> ppu.PrettyPrint:
   * nr_untapped: how many positional arguments (from the tail) should not be
   passed to the tap function.
   * arg_treedef: the treedef of the tapped positional arguments.
-  * transforms: a tuple of the transformations that have been applied.
-  * batch_dims: a tuple of the dims that have been batched, for vmap.
-  * logical_shapes: a tuple of evaluated logical shapes, for mask.
+  * transforms: a tuple of the transformations that have been applied. Most
+    elements of this tuple are a string naming the transformations. For some
+    transformation, the elements are a tuple of the name of the transformation
+    and the parameters. For example, for vmap, the parameters are the 
+    dimensions that have been batched. Also, for mask, the parameter is the 
+    logical shapes. 
 
   * the remaining parameters are passed to the tap function.
 """
-# TODO(necula): handle nested vmap and mask
 id_tap_p = core.Primitive("id_tap")
 id_tap_p.multiple_results = True
 xla.outfeed_primitives.add(id_tap_p)
 
-def _add_transform_name(params: Dict, transform: str) -> Dict:
+def _add_transform_name(params: Dict, transform: str,
+                        transform_params: Any = None) -> Dict:
   """Adds the `transform` to the params["transforms"]."""
-  return dict(params, transforms=params.get("transforms", ()) + (transform,))
+  if transform_params is None:
+    new_transform = transform
+  else:
+    new_transform = (transform, transform_params)
```

Comment:
This is a programmer's API (for `id_tap`), so I suspect homogeneous values of type `Tuple[str, Dict[str, Any]]` would be easiest to work with. If desired, it's easy to ignore an empty dict when processing values (e.g., in `id_print`).

Otherwise someone is likely to write a custom printing function that works when tested on auto-diff, but crashes when you try to use `vmap`.

Reply:
You are right, I will bite the bullet and make every entry in the transforms be a tuple.